core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -417,16 +418,19 @@ private[spark] object Utils extends Logging {
       input: FileChannel,
       output: WritableByteChannel,
       startPosition: Long,
-      bytesToCopy: Long): Unit = {
+      bytesToCopy: Long,
+      numTransferToCalls: Int): Unit = {
     val outputInitialState = output match {
       case outputFileChannel: FileChannel =>
         Some((outputFileChannel.position(), outputFileChannel))
       case _ => None
     }
     var count = 0L
+    var num = 0
     // In case transferTo method transferred less data than we have required.
-    while (count < bytesToCopy) {
+    while (count < bytesToCopy && num < numTransferToCalls) {
       count += input.transferTo(count + startPosition, bytesToCopy - count, output)
```

Comment:
I'd rather check if the return value of `input.transferTo` goes `0`, not temporally but consistently.

Reply:
Seems no big difference by checking `<= 0`, what's the advantage here?

Reply:
numTransferToCalls would depend on the size of bytesToCopy so harder to set which value makes sense to find the issue and do fail fast. Though `transferTo` can return 0 according to javadoc, if it would return 0 continuously (like 100 times) it sounds an issue to me.

We can also apply exponential wait here when the return value is 0, if it turns out calling transferTo right again doesn't help.

Reply:
While in a such infinite loop, it is quite fast to reach to a large number, say 10000 times of `transferTo` call. We don't need to set a small value, a large value like 1w+ is enough to cover most of the scenario and fail fast. 

Also it is not so good to add wait here, it will delay the shuffle write critical path, and in a such scenario adding delay will not recover anything, only just lower the system usage.

Reply:
I agree, the issue is making no progress at all right? Much better to check if nothing is being transferred for a long while. You can set a large value. Even if you reach it 'quickly', thousands of 0-byte transfers indicate a problem and you want to fail right?

The big big difference here is: say you cap the number of transfers at 10000, and you are trying to move 1GB, and it's just transferring 10K at a time. You would fail to write most of the data!

Reply:
Thanks @srowen for providing good example. That's exactly what I meant for first sentence.

> Also it is not so good to add wait here, it will delay the shuffle write critical path, and in a such scenario adding delay will not recover anything, only just lower the system usage.

If we assume IO will never be restored then yes maybe it doesn't help, but we don't have any evidence to back up. We're just logically thinking if it runs as an infinite loop, then the number should be 0 continuously. If we are suspecting this behavior, checking 0 should be closer to address our suspect, right? The advantage of checking `count` is, we are not restricted to check the count with `0`, as I'll also describe below.

If you're concerning about shuffle write as critical path and worry about latency, you may have some kind of SLA in mind (like N MiB/sec) and measuring this should provide more information to you. You may want to just kill it when you're writing huge shuffle but Spark writes shuffle equal or lower than 1 KiB/s for 10 mins - it'll write less than 1M in 10 mins and writing 1G of shuffle with such throughput is just effectively stuck. At least you may want to log message as debug or warn when the throughput for recent N mins doesn't make sense (I wouldn't say which number is good. Heuristic may come in, I don't know.).

We still get some hints from figuring out the relation of returned count and latency of `transferTo`. If it's tightly related, then we may have a chance to leverage the fact. If not, the elapsed time to finish the loop with loop count N is unpredictable and you may still waste minutes or even hours for loop count to be reaching configured number.

We will never reach a good answer unless there're some valuable numbers being measured during the error situation. Everything for now is just assumption.

Reply:
Alright, I will change to track the return value. But I don't have an evidence that the return value is always 0, it may be a very small value, I would fail it fast in a such scenario. 

As we tested on a normal cluster, 1 or 2 `transferTo` calls should be enough to copy more than 200MB data. If it cannot finish by calling more than 10k calls, I would fail it rather than waiting it.

Reply:
Checking `<=0` assumes that the return value is always 0, but if the return value is very small, shall we wait or make it fail? Normally a very small value means that disk/FS has some problems.

Reply:
Can you log the return of `transferTo` "locally" in some way to further test? Not sure how easy it is to reproduce. In the absence of evidence, I think checking <= 0 is the most that is safe. I think it bears understanding the issue more first before any change is made.

In any event I don't think you can ever just stop copying data and continue successfully. You would have to fail the job. 

If the job is merely slow, at best, you'd want to introduce some kind of timeout at a higher level as there are all kinds of ways it can be slow. But that would also take care of your current situation.

But you can always kill jobs that are slow directly, without any change to Spark.
This doesn't fix the cause anyway (assuming it's outside Spark).